File: csharp/protein-translation/ProteinTranslation.cs
using System.Collections.Generic;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        return Translate(strand).ToArray();
    }

    private static IEnumerable<string> Translate(string strand)
    {
        foreach (var codon in GetCodons(strand))
        {
            switch (codon)
            {
                case "AUG":
                    yield return "Methionine";
                    break;
                case "UUU":
                case "UUC":
                    yield return "Phenylalanine";
                    break;
                case "UUA":
                case "UUG":
                    yield return "Leucine";
                    break;
                case "UCU":
                case "UCC":
                case "UCA":
                case "UCG":
                    yield return "Serine";
                    break;
                case "UAU":
                case "UAC":
                    yield return "Tyrosine";
                    break;
                case "UGU":
                case "UGC":
                    yield return "Cysteine";
                    break;
                case "UGG":
                    yield return "Tryptophan";
                    break;
                case "UAA":
                case "UAG":
                case "UGA":
                    yield break;
            }
        }
    }

    private static IEnumerable<string>GetCodons(string strand)
    {
        for (int i = 0; i < strand.Length; i += 3)
        {
            yield return strand.Substring(i, 3);
        }

    }
}

File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let translate codon =
    match codon with
    | "AUG"                         -> Some("Methionine")
    | "UUU" | "UUC"                 -> Some("Phenylalanine")
    | "UUA" | "UUG"                 -> Some("Leucine")
    | "UCU" | "UCC" | "UCA" | "UCG" -> Some("Serine")
    | "UAU" | "UAC"                 -> Some("Tyrosine")
    | "UGU" | "UGC"                 -> Some("Cysteine")
    | "UGG"                         -> Some("Tryptophan")
    | "UAA" | "UAG" | "UGA"         -> None
    | _ -> invalidArg "codon" (codon + " is not a valid codon")

let proteins rna =
    rna
        |> Seq.chunkBySize 3
        |> Seq.map System.String
        |> Seq.map translate
        |> Seq.takeWhile (fun x -> x.IsSome)
        |> Seq.map (fun x -> x.Value)
        |> Seq.toList
